### PR TITLE
[xbar/dv] xbar exclusion automation

### PR DESCRIPTION
**Pull request by weicaiyang**
1. auto-generate xbar IP exclusion
2. disable VCS CHECK_SUM check for xbar
3. Delete old manually added exclusions
Will add automation for chip level xbar in next PR

Signed-off-by: Weicai Yang <weicai@google.com>

### CODE REVIEW

File: hw/ip/tlul/generic_dv/xbar_sim_cfg.hjson [hunk 1]
```diff
@@ -17,8 +17,8 @@
   // Fusesoc core file used for building the file list.
   fusesoc_core: "lowrisc:dv:{dut}_sim:0.1"
 
-  // Testplan hjson file: set in the autogenerated sim cfg file that imports this.
-  // testplan: ""
+  // Testplan hjson file.
+  testplan: "{proj_root}/hw/top_earlgrey/ip/{dut}/data/autogen/{dut}_testplan.hjson"
```
Comment (sriyerg) [inline hw/ip/tlul/generic_dv/xbar_sim_cfg.hjson:21]: I don't think its a good idea to put top level specific things in here. Any particular reason for moving these from the autogenerated sim cfg files?

File: hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+
+INSTANCE: tb.dut
+ANNOTATION: "[UNSUPPORTED]"
+Block 1 "0" "assign unused_scanmode = scanmode_i;"
+
+ANNOTATION: "[UNSUPPORTED]"
+Toggle scanmode_i
+
+ANNOTATION: "[UNR] these device address bits are always 0"
+Toggle tl_rom_o.a_address [14]
```
Comment (sriyerg) [inline hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el:31]: Range exclusion should also work here right? 
```suggestion
Toggle tl_rom_o.a_address [31:4]
```

File: hw/ip/tlul/generic_dv/xbar_sim_cfg.hjson (hunk 1 above)
Comment (weicaiyang) [reply hw/ip/tlul/generic_dv/xbar_sim_cfg.hjson:21]: they're the same for all IPs. If we need to support other than top_earlgrey, we probably need to update or add a variable like {top}.

File: hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el (hunk 2 above)
Comment (weicaiyang) [reply hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el:31]: I tried but EL file doesn't allow [31:4]
Spent a few hours to move toggle exclusion to cover.cfg. But looks like VCS doesn't support struct type in cover.cfg very well.
Thinking to just use EL file as it can locate the pin correctly even we use -excl_bypass_check. For the other exclusion, they're in stable modules, so it seems fine to bypass checksum 

Besides, the one we currently uses for top_earlgrey isn't actually applied. VCS can't find it with wildcard on struct.

> Warning-[VCM-HFUFR] Hier Config: regions not found
>   In the '-cm_hier' configuration file, pattern "+node --- 
>   tb.dut.top_earlgrey.u_*.tl_*" did not match any pattern.
>   Please check the entry at line number "16" in hier config file

**Comment by weicaiyang** [comment]
@sriyerg 
I added another commit in this PR to move toggle coverage exclusion automation to cover.cfg, which is more reliable.
the other coverage will remain in el file due to the limitation in cover.cfg.
I keep the first commit in case we feel it's better to just use EL file in the future.
PTAL

**Comment by weicaiyang** [comment]
@imphil 
I'm not sure why CI failed with following error.
I do have a license header for [xbar_cover.cfg.tpl](https://github.com/lowRISC/opentitan/blob/0bff64f631fd7850001bf8d542142426b69574bc/util/tlgen/xbar_cover.cfg.tpl). Can you take a look? Thanks
> ERROR: util/tlgen/xbar_cover.cfg.tpl:1 FAILED: File does not start with comment
> Results:      Files
> ----------  -------
> Passed           12
> Failed            1
> Skipped           2
> Excluded          0
> Total            15
> Failed:
>   util/tlgen/xbar_cover.cfg.tpl
> 
> ##[error]Licence header check failed. Please check output of util/lowrisc_misc-linters/licence-checker/licence-checker.py --
> config util/licence-checker.hjson on the noted failures.
> 
> ##[error]Bash exited with code '1'.
> Finishing: Check Licence Headers

**Comment by rswarbrick** [comment]
Hi @weicaiyang! The CI message shows a command that you can run to reproduce this locally. The problem is to do with licence headers in template files. It seems that the checker expects `foo.cfg.tpl` to use a `#` character to introduce a comment, not a `//`.

**Comment by weicaiyang** [comment]
> Hi @weicaiyang! The CI message shows a command that you can run to reproduce this locally. The problem is to do with licence headers in template files. It seems that the checker expects `foo.cfg.tpl` to use a `#` character to introduce a comment, not a `//`.

Thanks @rswarbrick. I updated `license_checker.py` to allow both comment styles.

**Comment by imphil** [comment]
The license checker is vendored in from https://github.com/lowRISC/misc-linters/ and you need to do a PR there. @lenary can help if needed.

**Comment by weicaiyang** [comment]
> The license checker is vendored in from https://github.com/lowRISC/misc-linters/ and you need to do a PR there. @lenary can help if needed.

@imphil @lenary 
I have updated https://github.com/lowRISC/misc-linters/, looks like the commit was just landed without a review.
How can I make this PR pass the CI? Do I need to vendor in that update rather than make a commit in this PR?

**Comment by imphil** [comment]
@weicaiyang you just pushed to the repository without a PR, which is not the way it is intended to be done, but not a problem at all in this case. To update the code in OpenTitan, you need to vendor the changes in. Run `./util/vendor.py ./util/lowrisc_misc-linters.vendor.hjson --commit --update` to vendor the latest changes in.

**Comment by weicaiyang** [comment]
> @weicaiyang you just pushed to the repository without a PR, which is not the way it is intended to be done, but not a problem at all in this case. To update the code in OpenTitan, you need to vendor the changes in. Run `./util/vendor.py ./util/lowrisc_misc-linters.vendor.hjson --commit --update` to vendor the latest changes in.

Thanks @imphil for the vendor-in command
I made this [PR](https://github.com/lowRISC/misc-linters/pull/13) to make isort pass. PTAL

File: util/tlgen/xbar_cov_excl.el.tpl [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+// xbar_cov_excl.el generated by `tlgen.py` tool
+
+ANNOTATION: "[NON_RTL]"
+MODULE: uvm_pkg
+Assert \uvm_reg_map::do_write .unnamed$$_0.unnamed$$_1 "assertion"
+Assert \uvm_reg_map::do_read .unnamed$$_0.unnamed$$_1 "assertion"
+
+ANNOTATION: "[UNSUPPORTED]"
```
Comment (sriyerg) [inline util/tlgen/xbar_cov_excl.el.tpl:12]: Could you please add more explanation for `unsupported`?

File: hw/ip/tlul/generic_dv/xbar_sim_cfg.hjson (hunk 1 above)
Comment (sriyerg) [reply hw/ip/tlul/generic_dv/xbar_sim_cfg.hjson:21]: Right, but it is easier to fix that in the template than here. Anyway, we can probably deal with it later.

File: hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cover.cfg [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+-module clk_rst_if  // DV construct.
+
+// due to VCS tool issue, can't move this part into begin...end (tgl) or after
+-node tb.dut tl_*.a_param
+-node tb.dut tl_*.d_param
+-node tb.dut tl_*.d_user
+-node tb.dut tl_*.a_user.*
+-node tb.dut tl_*.d_opcode[2:1]
+
+// [UNR] these device address bits are always 0"
```
Comment (sriyerg) [inline hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cover.cfg:18]:
```suggestion
// [UNR] these device address bits are always 0
```

File: hw/ip/tlul/generic_dv/xbar_sim_cfg.hjson [hunk 5]
```diff
@@ -17,14 +17,31 @@
...
+    }
+    {
+      name: default_vcs_cov_cfg_file
+      value: "-cm_hier {tool_srcs_dir}/xbar_cover.cfg"
+    }
+  ]
+
+  // Bypass VCS CHECK_SUM check as the exclusion file is generated without proper CHECK_SUM value
+  cov_analyze_opts: ["-excl_bypass_checks"]
+  cov_report_opts:  ["-excl_bypass_checks"]
```
Comment (sriyerg) [inline hw/ip/tlul/generic_dv/xbar_sim_cfg.hjson:40]: Hmm, these are VCS specific opts. Probably do something like this:
```suggestion
  vcs_cov_analyze_opts: ["-excl_bypass_checks"]
  vcs_cov_report_opts:  ["-excl_bypass_checks"]
  
  cov_analyze_opts: ["{tool}_cov_analyze_opts"]
  cov_report_opts: ["{tool}_cov_report_opts"]  
```

Need to find a better way to do things like this....

Comment (weicaiyang) [reply hw/ip/tlul/generic_dv/xbar_sim_cfg.hjson]: Updated, thanks

File: util/tlgen/xbar_cov_excl.el.tpl (hunk 3 above)
Comment (weicaiyang) [reply util/tlgen/xbar_cov_excl.el.tpl]: Added.

File: hw/ip/tlul/generic_dv/xbar_sim_cfg.hjson (hunk 1 above)
Comment (weicaiyang) [reply hw/ip/tlul/generic_dv/xbar_sim_cfg.hjson]: @sriyerg I reverted this back, this change can't support multiple top.

File: hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+
+Toggle tl_corei_i.a_param
+Toggle tl_corei_o.d_param
+
+Toggle tl_corei_o.d_user
+Toggle tl_corei_i.a_user.rsvd1
+Toggle tl_corei_i.a_user.parity
+Toggle tl_corei_i.a_user.parity_en
+
+Toggle tl_corei_o.d_opcode [2:1]
```
Comment (sriyerg) [inline hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el]: Does range selection work here? If yes then it should work on other fields of the struct as well, right?

File: hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el (hunk 2 above)
Comment (sriyerg) [reply hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el]: I think we should check with Will / file bug with Synopsys.

File: util/tlgen/xbar_cov_excl.el.tpl [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+MODULE: prim_arbiter_ppc ( parameter N=${i},DW=102,EnDataPort=1,EnReqStabA=0 ) 
+Condition 2 "175047464" "(valid_o && ((!ready_i))) 1 -1" (2 "10")
+%endfor
+ANNOTATION: "[NON_RTL]"
+MODULE: uvm_pkg
+Assert \uvm_reg_map::do_write .unnamed$$_0.unnamed$$_1 "assertion"
+Assert \uvm_reg_map::do_read .unnamed$$_0.unnamed$$_1 "assertion"
+
+INSTANCE: tb.dut
+ANNOTATION: "[UNSUPPORTED]"
```
Comment (sriyerg) [inline util/tlgen/xbar_cov_excl.el.tpl]: Flip these lines?

File: util/tlgen/xbar_cov_excl.el.tpl [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+Block 1 "0" "assign unused_scanmode = scanmode_i;"
+
+ANNOTATION: "[UNSUPPORTED]"
+Toggle scanmode_i
+
+ANNOTATION: "[UNR] these device address bits are always 0"
```
Comment (sriyerg) [inline util/tlgen/xbar_cov_excl.el.tpl]:
```suggestion
ANNOTATION: "[UNR] these device address bits are always 0
```

File: hw/ip/tlul/generic_dv/xbar_tests.hjson [hunk 9]
```diff
@@ -125,6 +125,8 @@
       en_run_modes: ["xbar_mode"]
       uvm_test: xbar_error_test
       uvm_test_seq: xbar_random_vseq
+      // More seeds to cover assert coverage - d_errorChangedNotAccepted_C
+      reseed: 100
```
Comment (sriyerg) [inline hw/ip/tlul/generic_dv/xbar_tests.hjson:129]: Can the constraints in the sequence be adjusted to hit this a bit more favorably rather than increase seeds? I am ok if that is too hard.

File: hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cover.cfg [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// xbar_cover.cfg generated by `tlgen.py` tool
+
++tree tb.dut
+-module pins_if     // DV construct.
+-module clk_rst_if  // DV construct.
+
+// due to VCS tool issue, can't move this part into begin...end (tgl) or after
```
Comment (sriyerg) [inline hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cover.cfg:11]: This is fixed in VCS version 2020.12.
https://solvnetplus.synopsys.com/s/detail/a0f3g000000U5cD

File: hw/top_earlgrey/ip/xbar_peri/dv/autogen/xbar_cov_excl.el [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+Toggle scanmode_i
+
+ANNOTATION: "[UNR] these device address bits are always 0"
+Toggle tl_uart_o.a_address [12]
+Toggle tl_uart_o.a_address [13]
+Toggle tl_uart_o.a_address [14]
+Toggle tl_uart_o.a_address [15]
+Toggle tl_uart_o.a_address [16]
+Toggle tl_uart_o.a_address [17]
+Toggle tl_uart_o.a_address [18]
```
Comment (sriyerg) [inline hw/top_earlgrey/ip/xbar_peri/dv/autogen/xbar_cov_excl.el]: If these are added to `cover.cfg` then it can be removed from here right?

File: util/dvsim/Modes.py [hunk 12]
```diff
@@ -118,18 +118,19 @@ def merge_mode(self, mode):
             if type(self_attr_val) != type(mode_attr_val):
                 log.error(
                     "Mode %s cannot be merged into %s due to a conflict "
-                    "(type mismatch): %s: {%s(%s), %s(%s)}", name, self.name,
-                    attr, str(self_attr_val), str(type(self_attr_val)),
-                    str(mode_attr_val), str(type(mode_attr_val)))
+                    "(type mismatch): %s: {%s(%s), %s(%s)}", mode.name,
```
Comment (sriyerg) [inline util/dvsim/Modes.py:121]: This fix is not related to XBAR. It must be in its own commit.

File: hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el (hunk 2 above)
Comment (weicaiyang) [reply hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el]: Actually I already asked and know how to fix it. Will make a PR to fix the top-level

File: hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_main_sim_cfg.hjson [hunk 13]
```diff
@@ -12,9 +12,6 @@
   // Testplan hjson file.
   testplan: "{proj_root}/hw/top_earlgrey/ip/{dut}/data/autogen/{dut}_testplan.hjson"
 
-  // Add xbar_main specific exclusion files.
```
Comment (sriyerg) [inline hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_main_sim_cfg.hjson:15]: How are the auto-generated cover cfg and el file set in the HJson?

Comment (weicaiyang) [reply hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_main_sim_cfg.hjson:15]: Good catch. I accidentally removed this when resolving the conflict. Updated, PTAL

File: hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// xbar_cov_excl.el generated by `tlgen.py` tool
+
+ANNOTATION: "[NON_RTL]"
+MODULE: uvm_pkg
+Assert \uvm_reg_map::do_write .unnamed$$_0.unnamed$$_1 "assertion"
+Assert \uvm_reg_map::do_read .unnamed$$_0.unnamed$$_1 "assertion"
+
```
Comment (sriyerg) [inline hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el:11]: minor nit: These new lines are a bit arbitrary. Probably get rid of them (or add one before each ANNOTATION line).

**Review by sriyerg — APPROVED** [review]
LGTM, thanks!

File: hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el (hunk 14 above)
Comment (weicaiyang) [reply hw/top_earlgrey/ip/xbar_main/dv/autogen/xbar_cov_excl.el:11]: Those with no new lines are generated in a loop. Added a new line in the loop as well. Thanks
